### PR TITLE
feature: HollowHistory can build history using reverse deltas to allow building history in both directions simultaneously

### PR DESCRIPTION
**Pull request by Sunjeet**
Peering credit to @shoc2020 

HollowHistory maintains two read states- one that always moves in the forward direction and one that always moves in the reverse direction. Once max retention has been reached the read state corresponding to the reverse direction is dropped and (as before) each future forward delta application results in eviction of the oldest historical state.

Building history in reverse allows recent data versions to be available in the history tooling sooner at the time of HollowHistory initialization, and also has the nice property that history initialization uses the latest schema at the time of initialization so recently introduced backwards compatible schema changes are visible without the need to apply double snapshots.

The UI view of History built using reverse deltas is at parity with when building using forward deltas. 

NOTE: if computing history using reverse deltas, for header tags to be displayed accurately the producer needs to be on Hollow v7.0.2 - that release fixed a bug in how headers tags were set on reverse deltas.

Example usage:
(continue existing usage for fwd only)
```
consumerFwd.triggerRefreshTo(v);
HollowHistoryUIServer historyUIServer = new HollowHistoryUIServer(consumerFwd, PORT);
historyUIServer.start();
consumerFwd.triggerRefreshTo(v+1);
consumerFwd.triggerRefreshTo(v+2);
...
```
(new for building history in both directions)
```
consumerFwd.triggerRefreshTo(v);
  consumerRev.triggerRefreshTo(v);
HollowHistoryUIServer historyUIServer = new HollowHistoryUIServer(consumerFwd, consumerRev, PORT);
historyUIServer.start();
consumerFwd.triggerRefreshTo(v+1);
  consumerRev.triggerRefreshTo(v-1);
  consumerRev.triggerRefreshTo(v-2);
  consumerRev.triggerRefreshTo(v-3);
consumerFwd.triggerRefreshTo(v+2);
...

```

### CODE REVIEW

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryRefreshListener.java
Comment (mbrilnetflix) [inline hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryRefreshListener.java:29]: The grammar of this sentence seems off

Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryRefreshListener.java:29]: reworded

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (mbrilnetflix) [inline hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: nit: Please add comment to clarify this will be a Forward only

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (mbrilnetflix) [inline hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: we don't support a rev only consumer? in that case, consumerFwd will be null and throw a NPE
If we do not, it's fine, we can add a null check.

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (mbrilnetflix) [inline hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: NPE if Rev is null

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java
Comment (mbrilnetflix) [inline hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java]: checkstyle didn't complain about this?

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java
Comment (mbrilnetflix) [inline hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java:172]: this can be confusing IMHO, 
We should throw exception if either is set and other is not.

File: hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowHistoryTypeKeyIndex.java
Comment (mbrilnetflix) [inline hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowHistoryTypeKeyIndex.java]: nit: this seems odd fomating

File: hollow/src/test/java/com/netflix/hollow/core/util/RemovedOrdinalIteratorTest.java
Comment (mbrilnetflix) [inline hollow/src/test/java/com/netflix/hollow/core/util/RemovedOrdinalIteratorTest.java]: these are not reset on each test (when multiple test cases are ran)

**Review by mbrilnetflix — COMMENTED** [review]
LGTM if assumptions mentioned in my comments are confirmed, 
I'll defer to Drew for this one, as I don't have full context on this.

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java
Comment (Sunjeet) [inline hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java:572]: Ah, just learned that `HollowHistoricalState`s hold some references into the typestates contained within `oldestHollowHistoryState` so dereferencing it here won't actually free up much memory. I think we can drop those references once the `HollowHistoricalState` has been populated. Adding that...

Comment (Sunjeet) [reply hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java:572]: Added code to dereference read state engine once HollowHistoricalState has been populated so it can be GC'ed.

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: The underlying `HollowHistory` initialization has a 
```
requireNonNull(fwdMovingHollowReadStateEngine, "Fwd direction read state engine should be initialized");
```
and yes, we don't support rev only because it isn't really useful and it would add complexity- api that accepts a single state engine would need to also accept/determine which direction it will be traversing in

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: consumerRev is null checked before getting into this `else` block. ConsumerFwd is null checked within HollowHistory init

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java
Comment (Sunjeet) [reply hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java:172]: yes! the call to `initializeReverseStateEngine` in the next line validates the inputs

File: hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowHistoryTypeKeyIndex.java
Comment (Sunjeet) [reply hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowHistoryTypeKeyIndex.java]: fixed

File: hollow/src/test/java/com/netflix/hollow/core/util/RemovedOrdinalIteratorTest.java
Comment (Sunjeet) [reply hollow/src/test/java/com/netflix/hollow/core/util/RemovedOrdinalIteratorTest.java]: They weren't intended to be. I failed to make them static final. Now addressed.

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (dkoszewnik) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: +1

Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: added

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java
Comment (Sunjeet) [reply hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java]: Sounds like something python linter would do :) 
Anyway, fixed

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistoricalStateCreator.java
Comment (dkoszewnik) [inline hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistoricalStateCreator.java:111]: Can you please help me understand why we need this?

**Review by dkoszewnik — APPROVED** [review]
It looks fantastic and solves an important problem.  I'm very excited that our history/explorer UI will soon detect schema changes, then re-initialize the history in reverse, so new fields will be automatically visible while still also providing visibility into a long history of changes.

File: hollow/src/main/java/com/netflix/hollow/core/read/engine/list/HollowListDeltaHistoricalStateCreator.java
Comment (nayanika-u) [inline hollow/src/main/java/com/netflix/hollow/core/read/engine/list/HollowListDeltaHistoricalStateCreator.java:73]: A comment or java doc would be helpful here to understand why/where this method is required.

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (nayanika-u) [inline hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java:96]: nit : could we make the max states configurable?

File: hollow-diff-ui/src/test/java/com/netflix/hollow/diffview/HollowHistoryUITest.java
Comment (nayanika-u) [inline hollow-diff-ui/src/test/java/com/netflix/hollow/diffview/HollowHistoryUITest.java:223]: How does the HistoryUIServer recover in the 2 scenarios - a) revConsumer and fwdConsumer are on different versions at init - When would that happen? b) revConsumer not initialized. Do we restart the history server with less or maybe 1 state?

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java]: (on the constructor)

File: hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistoricalStateCreator.java
Comment (Sunjeet) [reply hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistoricalStateCreator.java:111]: Added code comment

File: hollow/src/main/java/com/netflix/hollow/core/read/engine/list/HollowListDeltaHistoricalStateCreator.java
Comment (Sunjeet) [reply hollow/src/main/java/com/netflix/hollow/core/read/engine/list/HollowListDeltaHistoricalStateCreator.java:73]: Added

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java:96]: Yup they are, this constructor is just plugging in a default to the other constructor that accepts it as a param

File: hollow-diff-ui/src/test/java/com/netflix/hollow/diffview/HollowHistoryUITest.java
Comment (Sunjeet) [reply hollow-diff-ui/src/test/java/com/netflix/hollow/diffview/HollowHistoryUITest.java:223]: (a) The history server fails hard with an IllegalStateException at initialization when it detects that reverse consumer didnt start at the same version that the fwd consumer started with. There is also a unit test `historyUsingFwdAndRevConsumer_revAndFwdConsumersMustBeOnSameVersionAtInit` in `HollowHistoryUITest.java` that exercises this scenario. We could potentially make improvements to handle the various initialization scenarios like there is a gap in the start versions that gets stitched together,  or the fwd and rev initialize so that they will pass each other at some point. There isnt a compelling reason to handle those scenarios at the moment.

Comment (Sunjeet) [reply hollow-diff-ui/src/test/java/com/netflix/hollow/diffview/HollowHistoryUITest.java:223]: (b) revConsumer can be initialized at any point (for e.g. after >1 historical states have been created using the fwd consumer). So if fwd and rev start at v0, at that point there are no historical states. When fwd transitions to v1 we have one historical state (historical state v1 representing the diff between v0 and v1). More fwd transition applications would lead to more historical states v2, v3, v4, etc. When the rev consumer comes along its first transition (v0 to v-1) will result in the creation of the historical state v0 (giving the same effect as if the fwd consumer had transitioned from v-1 to v0).

File: hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java
Comment (Sunjeet) [reply hollow-diff-ui/src/main/java/com/netflix/hollow/history/ui/HollowHistoryUI.java:96]: Hah, we were missing the max states configurable flavor constructor in the `HollowHistoryUIServer` class. Not sure if thats the one you may have been thinking of here. Now added- https://github.com/Netflix/hollow/commit/d4d3dbed944a8885ba128843a61e569ddbcce08e

**Pull request merged**
